indexer/services/comlink/src/controllers/api/v4/compliance-v2-controller.ts

```diff
@@ -335,6 +276,161 @@ function generateAddress(pubkeyArray: Uint8Array): string {
+  const now = DateTime.now().toSeconds();
+  if (Math.abs(now - timestamp) > GEOBLOCK_REQUEST_TTL_SECONDS) {
+    return create4xxResponse(
+      res,
+      `Timestamp is not within the valid range of ${GEOBLOCK_REQUEST_TTL_SECONDS} seconds`,
+    );
+  }
+
+  // Prepare the message for verification
+  const messageToSign: string = `${message}:${action}"${currentStatus || ''}:${timestamp}`;
+  const messageHash: Uint8Array = sha256(Buffer.from(messageToSign));
+  const signedMessageArray: Uint8Array = new Uint8Array(Buffer.from(signedMessage, 'base64'));
+  const signature: ExtendedSecp256k1Signature = ExtendedSecp256k1Signature
+    .fromFixedLength(signedMessageArray);
+
+  // Verify the signature
+  const isValidSignature: boolean = await Secp256k1.verifySignature(
+    signature,
+    messageHash,
+    pubkeyArray,
+  );
+  if (!isValidSignature) {
+    return create4xxResponse(
+      res,
+      'Signature verification failed',
+    );
+  }
+
+  return undefined;
+}
+
+function validateSignatureKeplr(
+  res:express.Response,
+  address: string,
+  message: string,
+  signedMessage: string,
+  pubkey: string,
+): express.Response| undefined {
+  const messageToSign: string = message;
+
+  const pubKeyUint = new Uint8Array(Buffer.from(pubkey, 'base64'));
+  const signedMessageUint = new Uint8Array(Buffer.from(signedMessage, 'base64'));
+
+  const isVerified = verifyADR36Amino('dydx', address, messageToSign, pubKeyUint, signedMessageUint, 'secp256k1');
+
+  if (!isVerified) {
+    return create4xxResponse(
+      res,
+      'Keplr signature verification failed',
+    );
+  }
+
+  return undefined;
+}
+
+async function checkCompliance(
+  req: express.Request, res: express.Response, address: string, action: ComplianceAction,
+): Promise<express.Response> {
+  if (isWhitelistedAddress(address)) {
+    return res.send({
+      status: ComplianceStatus.COMPLIANT,
+      updatedAt: DateTime.utc().toISO(),
+    });
+  }
+
+  const [
+    complianceStatus,
+    wallet,
+  ]: [
+    ComplianceStatusFromDatabase[],
+    WalletFromDatabase | undefined,
+  ] = await Promise.all([
+    ComplianceStatusTable.findAll(
+      { address: [address] },
+      [],
+    ),
+    WalletTable.findById(address),
+  ]);
+
+  const updatedAt: string = DateTime.utc().toISO();
+  const complianceStatusFromDatabase:
+  ComplianceStatusFromDatabase | undefined = await upsertComplianceStatus(
+    req,
+    action,
+    address,
+    wallet,
+    complianceStatus,
+    updatedAt,
+  );
+  if (complianceStatus.length === 0 ||
+    complianceStatus[0] !== complianceStatusFromDatabase) {
+    if (complianceStatusFromDatabase !== undefined &&
+      complianceStatusFromDatabase.status !== ComplianceStatus.COMPLIANT
+    ) {
+      stats.increment(
+        `${config.SERVICE_NAME}.${controllerName}.geo_block.compliance_status_changed.count`,
+        {
+          newStatus: complianceStatusFromDatabase!.status,
+        },
```

Comment:
nit: could you also add if this was a regular or a Keplr signature in teh stats?